File: src/soundboard/manager.py
import json
import logging
import os
import shutil
import uuid
from dataclasses import asdict, dataclass
from typing import Dict, List, Optional
from .player import SoundboardPlayer, SoundTrack

logger = logging.getLogger("Audiover.SoundboardManager")


@dataclass
class SoundItem:
    id: str
    name: str
    file_path: str
    hotkey: Optional[str] = None
    volume: float = 1.0
    loop: bool = False
    category: str = "General"
    duration_sec: float = 0.0


class SoundboardManager:
    """Manages soundboard metadata, library organization, and settings persistence."""

    def __init__(
        self,
        config_path: str = "config/settings.json",
        sounds_dir: str = "assets/sounds",
        player: Optional[SoundboardPlayer] = None,
    ):
        self.config_path = config_path
        self.sounds_dir = sounds_dir
        self.player = player or SoundboardPlayer()
        self.sounds: Dict[str, SoundItem] = {}

        os.makedirs(self.sounds_dir, exist_ok=True)
        os.makedirs(os.path.dirname(self.config_path), exist_ok=True)

    def load_from_config(self):
        """Loads configured sounds from settings.json and pre-loads audio into player."""
        if not os.path.exists(self.config_path):
            return

        try:
            with open(self.config_path, "r", encoding="utf-8") as f:
                data = json.load(f)

            sound_list = data.get("soundboard", {}).get("sounds", [])
            for item in sound_list:
                s_id = item.get("id", str(uuid.uuid4()))
                file_path = item.get("file_path", "")
                if not os.path.exists(file_path):
                    # Check relative path inside assets/sounds
                    alt_path = os.path.join(
                        self.sounds_dir, os.path.basename(file_path)
                    )
                    if os.path.exists(alt_path):
                        file_path = alt_path
                    else:
                        logger.warning(
                            f"Sound file not found: {file_path}, skipping..."
                        )
                        continue

                sound_item = SoundItem(
                    id=s_id,
                    name=item.get(
                        "name",
                        os.path.splitext(os.path.basename(file_path))[0],
                    ),
                    file_path=file_path,
                    hotkey=item.get("hotkey"),
                    volume=float(item.get("volume", 1.0)),
                    loop=bool(item.get("loop", False)),
                    category=item.get("category", "General"),
                    duration_sec=float(item.get("duration_sec", 0.0)),
                )

                # Preload into player
                track = self.player.load_sound(
                    sound_id=s_id,
                    file_path=file_path,
                    name=sound_item.name,
                    volume=sound_item.volume,
                    loop=sound_item.loop,
                )
                if track:
                    sound_item.duration_sec = track.duration_sec
                    self.sounds[s_id] = sound_item

            logger.info(f"Loaded {len(self.sounds)} sounds into soundboard.")
        except Exception as e:
            logger.error(f"Error loading soundboard config: {e}")

    def save_to_config(self):
        """Persists soundboard metadata into settings.json."""
        try:
            settings = {}
            if os.path.exists(self.config_path):
                with open(self.config_path, "r", encoding="utf-8") as f:
                    settings = json.load(f)

            if "soundboard" not in settings:
                settings["soundboard"] = {}

            settings["soundboard"]["sounds"] = [
                asdict(item) for item in self.sounds.values()
            ]

            with open(self.config_path, "w", encoding="utf-8") as f:
                json.dump(settings, f, indent=2, ensure_ascii=False)
            logger.info("Saved soundboard configuration to settings.json.")
        except Exception as e:
            logger.error(f"Failed to save soundboard config: {e}")

    def add_sound_file(
        self,
        file_path: str,
        name: Optional[str] = None,
        copy_to_assets: bool = False,
        hotkey: Optional[str] = None,
        volume: float = 1.0,
        loop: bool = False,
        category: str = "General",
    ) -> Optional[SoundItem]:
        """Adds a sound to the library and loads it into memory."""
        if not os.path.exists(file_path):
            logger.error(f"File not found: {file_path}")
            return None

        sound_id = str(uuid.uuid4())[:8]
        if not name:
            name = os.path.splitext(os.path.basename(file_path))[0]

        target_path = file_path
        if copy_to_assets:
            dest_filename = f"{sound_id}_{os.path.basename(file_path)}"
            target_path = os.path.join(self.sounds_dir, dest_filename)
            shutil.copy2(file_path, target_path)

        track = self.player.load_sound(
            sound_id=sound_id,
            file_path=target_path,
            name=name,
            volume=volume,
            loop=loop,
        )

        if not track:
            return None

        sound_item = SoundItem(
            id=sound_id,
            name=name,
            file_path=target_path,
            hotkey=hotkey,
            volume=volume,
            loop=loop,
            category=category,
            duration_sec=track.duration_sec,
        )
        self.sounds[sound_id] = sound_item
        self.save_to_config()
        return sound_item

    def remove_sound(self, sound_id: str):
        """Removes a sound from library and player."""
        if sound_id in self.sounds:
            self.player.remove_sound(sound_id)
            del self.sounds[sound_id]
            self.save_to_config()

    def update_sound(
        self,
        sound_id: str,
        name: Optional[str] = None,
        hotkey: Optional[str] = None,
        volume: Optional[float] = None,
        loop: Optional[bool] = None,
    ):
        """Updates metadata and syncs with player."""
        item = self.sounds.get(sound_id)
        if not item:
            return

        if name is not None:
            item.name = name
        if hotkey is not None:
            item.hotkey = hotkey if hotkey.strip() else None
        if volume is not None:
            item.volume = volume
            self.player.set_volume(sound_id, volume)
        if loop is not None:
            item.loop = loop
            self.player.set_loop(sound_id, loop)

        self.save_to_config()

    def get_sound(self, sound_id: str) -> Optional[SoundItem]:
        return self.sounds.get(sound_id)

    def get_all_sounds(self) -> List[SoundItem]:
        return list(self.sounds.values())

File: src/src/soundboard/manager.rs
use super::player::SoundboardPlayer;
use log::{error, info, warn};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::fs;
use std::path::{Path, PathBuf};
use std::sync::Arc;
use uuid::Uuid;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SoundItem {
    pub id: String,
    pub name: String,
    pub file_path: String,
    #[serde(default)]
    pub hotkey: Option<String>,
    #[serde(default = "default_volume")]
    pub volume: f32,
    #[serde(default)]
    pub loop_playback: bool,
    #[serde(default = "default_category")]
    pub category: String,
    #[serde(default)]
    pub duration_sec: f32,
}

fn default_volume() -> f32 {
    1.0
}

fn default_category() -> String {
    "General".to_string()
}

pub struct SoundboardManager {
    pub config_path: PathBuf,
    pub sounds_dir: PathBuf,
    pub player: Arc<SoundboardPlayer>,
    pub sounds: parking_lot::RwLock<HashMap<String, SoundItem>>,
}

impl SoundboardManager {
    pub fn new(
        config_path: PathBuf,
        sounds_dir: PathBuf,
        player: Arc<SoundboardPlayer>,
    ) -> Self {
        let _ = fs::create_dir_all(&sounds_dir);
        if let Some(parent) = config_path.parent() {
            let _ = fs::create_dir_all(parent);
        }

        Self {
            config_path,
            sounds_dir,
            player,
            sounds: parking_lot::RwLock::new(HashMap::new()),
        }
    }

    pub fn load_from_config(&self) {
        if !self.config_path.exists() {
            return;
        }

        let content = match fs::read_to_string(&self.config_path) {
            Ok(c) => c,
            Err(e) => {
                error!("Could not read config file: {}", e);
                return;
            }
        };

        let json_val: serde_json::Value = match serde_json::from_str(&content) {
            Ok(v) => v,
            Err(e) => {
                error!("Failed to parse config json: {}", e);
                return;
            }
        };

        let sounds_arr = json_val
            .get("soundboard")
            .and_then(|sb| sb.get("sounds"))
            .and_then(|s| s.as_array());

        if let Some(sound_list) = sounds_arr {
            let mut map = self.sounds.write();
            for item_val in sound_list {
                let id = item_val
                    .get("id")
                    .and_then(|v| v.as_str())
                    .map(|s| s.to_string())
                    .unwrap_or_else(|| Uuid::new_v4().to_string()[..8].to_string());

                let mut file_path = item_val
                    .get("file_path")
                    .and_then(|v| v.as_str())
                    .unwrap_or_default()
                    .to_string();

                if !Path::new(&file_path).exists() {
                    let alt = self.sounds_dir.join(
                        Path::new(&file_path)
                            .file_name()
                            .unwrap_or_default(),
                    );
                    if alt.exists() {
                        file_path = alt.to_string_lossy().to_string();
                    } else {
                        warn!("Sound file not found: {}, skipping...", file_path);
                        continue;
                    }
                }

                let name = item_val
                    .get("name")
                    .and_then(|v| v.as_str())
                    .map(|s| s.to_string())
                    .unwrap_or_else(|| {
                        Path::new(&file_path)
                            .file_stem()
                            .unwrap_or_default()
                            .to_string_lossy()
                            .to_string()
                    });

                let hotkey = item_val
                    .get("hotkey")
                    .and_then(|v| v.as_str())
                    .map(|s| s.to_string());

                let volume = item_val
                    .get("volume")
                    .and_then(|v| v.as_f64())
                    .map(|f| f as f32)
                    .unwrap_or(1.0);

                let loop_playback = item_val
                    .get("loop")
                    .and_then(|v| v.as_bool())
                    .unwrap_or(false);

                let category = item_val
                    .get("category")
                    .and_then(|v| v.as_str())
                    .unwrap_or("General")
                    .to_string();

                let mut duration_sec = item_val
                    .get("duration_sec")
                    .and_then(|v| v.as_f64())
                    .map(|f| f as f32)
                    .unwrap_or(0.0);

                if let Some(track) = self.player.load_sound(
                    &id,
                    &file_path,
                    Some(&name),
                    volume,
                    loop_playback,
                ) {
                    duration_sec = track.duration_sec;
                }

                let sound_item = SoundItem {
                    id: id.clone(),
                    name,
                    file_path,
                    hotkey,
                    volume,
                    loop_playback,
                    category,
                    duration_sec,
                };

                map.insert(id, sound_item);
            }
            info!("Loaded {} sounds into soundboard manager.", map.len());
        }
    }

    pub fn save_to_config(&self) {
        let mut settings: serde_json::Value = if self.config_path.exists() {
            fs::read_to_string(&self.config_path)
                .ok()
                .and_then(|c| serde_json::from_str(&c).ok())
                .unwrap_or_else(|| serde_json::json!({}))
        } else {
            serde_json::json!({})
        };

        let sounds_map = self.sounds.read();
        let sound_items_vec: Vec<serde_json::Value> = sounds_map
            .values()
            .map(|item| {
                serde_json::json!({
                    "id": item.id,
                    "name": item.name,
                    "file_path": item.file_path,
                    "hotkey": item.hotkey,
                    "volume": item.volume,
                    "loop": item.loop_playback,
                    "category": item.category,
                    "duration_sec": item.duration_sec
                })
            })
            .collect();

        if !settings.is_object() {
            settings = serde_json::json!({});
        }

        if settings.get("soundboard").is_none() {
            settings["soundboard"] = serde_json::json!({});
        }
        settings["soundboard"]["sounds"] = serde_json::json!(sound_items_vec);

        if let Ok(formatted) = serde_json::to_string_pretty(&settings) {
            let _ = fs::write(&self.config_path, formatted);
            info!("Saved soundboard settings to {:?}", self.config_path);
        }
    }

    pub fn add_sound_file(
        &self,
        file_path: &str,
        name: Option<&str>,
        copy_to_assets: bool,
        hotkey: Option<String>,
        volume: f32,
        loop_playback: bool,
    ) -> Option<SoundItem> {
        let path = Path::new(file_path);
        if !path.exists() {
            return None;
        }

        let sound_id = Uuid::new_v4().to_string()[..8].to_string();
        let final_name = name
            .map(|s| s.to_string())
            .unwrap_or_else(|| path.file_stem().unwrap_or_default().to_string_lossy().to_string());

        let target_path = if copy_to_assets {
            let dest_name = format!("{}_{}", sound_id, path.file_name().unwrap_or_default().to_string_lossy());
            let dest = self.sounds_dir.join(dest_name);
            let _ = fs::copy(path, &dest);
            dest.to_string_lossy().to_string()
        } else {
            file_path.to_string()
        };

        let track = self.player.load_sound(
            &sound_id,
            &target_path,
            Some(&final_name),
            volume,
            loop_playback,
        )?;

        let item = SoundItem {
            id: sound_id.clone(),
            name: final_name,
            file_path: target_path,
            hotkey,
            volume,
            loop_playback,
            category: "General".to_string(),
            duration_sec: track.duration_sec,
        };

        self.sounds.write().insert(sound_id, item.clone());
        self.save_to_config();
        Some(item)
    }

    pub fn update_sound(
        &self,
        id: &str,
        volume: Option<f32>,
        loop_playback: Option<bool>,
        hotkey: Option<String>,
    ) -> bool {
        let mut map = self.sounds.write();
        if let Some(item) = map.get_mut(id) {
            if let Some(v) = volume {
                item.volume = v;
            }
            if let Some(lp) = loop_playback {
                item.loop_playback = lp;
            }
            if let Some(hk) = hotkey {
                item.hotkey = if hk.trim().is_empty() { None } else { Some(hk) };
            }
            self.player.update_track(id, volume, loop_playback);
            drop(map);
            self.save_to_config();
            true
        } else {
            false
        }
    }

    pub fn remove_sound(&self, id: &str) -> bool {
        let mut map = self.sounds.write();
        if let Some(item) = map.remove(id) {
            self.player.remove_track(id);
            // If in sounds_dir, delete file
            let p = Path::new(&item.file_path);
            if p.starts_with(&self.sounds_dir) && p.exists() {
                let _ = fs::remove_file(p);
            }
            drop(map);
            self.save_to_config();
            true
        } else {
            false
        }
    }

    pub fn get_all_sounds(&self) -> Vec<SoundItem> {
        self.sounds.read().values().cloned().collect()
    }

    pub fn play_by_hotkey(&self, key: &str) -> bool {
        let normalized = key.trim().to_uppercase();
        let map = self.sounds.read();
        for sound in map.values() {
            if let Some(hk) = &sound.hotkey {
                if hk.trim().to_uppercase() == normalized {
                    let id = sound.id.clone();
                    drop(map);
                    if self.player.is_playing(&id) {
                        self.player.stop(&id);
                    } else {
                        self.player.play(&id);
                    }
                    return true;
                }
            }
        }
        false
    }
}
